### ISSUE TITLE
Duplicate variable shown in Tensorboard expected?

### ISSUE THREAD

**Issue by ethanluoyc** [tensorflow/tensorflow#9545]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
MacOS Sierra 12.12.4
- **TensorFlow installed from (source or binary)**:
pip
- **TensorFlow version (use command below)**:
1.1.0 (CPU)
### Describe the problem
I am trying to implement E2C (available from https://arxiv.org/pdf/1506.07365.pdf). Basically it is a neural network that is used for learning a transition model using neural networks. In the training set I have data of the form (X_t, X_t+1) where both X_t and X_t+1 needs to be transformed by an encoding network (e.g. a variational autoencoder). I use the following snippet for creating the encoding network (adapted from https://github.com/ericjang/e2c):

```python
    def encode(self, x, share=None):
        fc = tf.contrib.layers.fully_connected
        with tf.variable_scope('Encoder', reuse=share):
            l1 = fc(x, 400, weights_initializer=tf.orthogonal_initializer(),
                    activation_fn=tf.nn.relu)
            l2 = fc(l1, 100, weights_initializer=tf.orthogonal_initializer(),
                    activation_fn=tf.nn.relu)
            return l2

    def decode(self, z, share=None):
        fc = tf.contrib.layers.fully_connected
        with tf.variable_scope("Decoder", reuse=share):
            l1 = fc(z, 100, weights_initializer=tf.orthogonal_initializer(1.1),
                    activation_fn=tf.nn.relu)
            l2 = fc(l1, 400, weights_initializer=tf.orthogonal_initializer(1.1),
                    activation_fn=tf.nn.relu)

            return fc(l2, self.x_dim,
                      weights_initializer=tf.orthogonal_initializer(1.1),
                      activation_fn=tf.nn.sigmoid)
```
Then I would use something like

```python
h_enc_t = encoder(X_t)
h_enc_t_next = encoder(X_{t+1}, share=True)
```
to create the encoded output for the model.

The problem is that when visualizing this on Tensorboard, while it is sharing the variables by setting `share=True` for the variable scope, on the graph visulisation you will have `Encoder` and `Encoder_1` instead of just a `Decoder` scope. Of course they took different input since we need to transform X_t and X_t+1, but shouldn't the network be wrapped in the same scope since underneath we are reusing the same weights? I wonder if it is a feature to have `Encoder_1` and `Encoder` separately or it is a limitation of the variable scoping. The problem is illustrated in the screenshot below, you will see duplicates for 'Encoder' 'SampleQPhi" etc:

![graph-run](https://cloud.githubusercontent.com/assets/6040760/25559064/98fecc5a-2d2b-11e7-8669-00b1227abf17.png)

However, I would expect something like this (as appeared in the paper) to be a more reasonable visualization (h_enc) with input x_t and x_t+1 are the same network:

<img width="546" alt="screenshot 2017-04-30 15 51 01" src="https://cloud.githubusercontent.com/assets/6040760/25565322/fc04ab8a-2dbc-11e7-9876-5e823a3bf47b.png">

Many thanks in advance!

**Comment by aselle** [comment]
@dandelionmane, could you please take a look?

**Comment by chihuahua** [comment]
I need to take a look more closely, but I think calling `tf.scope` creates a new namespace in general, appending _[Some int] to the name if necessary for uniqueness.

This might deviate from your intention, but what if you move the `Encoder` scope outside of `encode` like this?

```
with tf.variable_scope('Encoder'):
  h_enc_t = encoder(X_t)
  h_enc_t_next = encoder(X_{t+1}, share=True)
```

And then

```
def encode(self, x, share=None):
  fc = tf.contrib.layers.fully_connected
  with tf.variable_scope('encode', reuse=share):
    l1 = fc(x, 400, weights_initializer=tf.orthogonal_initializer(),
            activation_fn=tf.nn.relu)
    l2 = fc(l1, 100, weights_initializer=tf.orthogonal_initializer(),
            activation_fn=tf.nn.relu)
    return l2
```

?

**Comment by ethanluoyc** [comment]
@chihuahua Nope this won't work. You would obtain something that's even weirder.

![image](https://cloud.githubusercontent.com/assets/6040760/25756118/81e85bee-31bd-11e7-9ad3-016893b18d14.png)

The reason I think this is not the intended behavior is because. In some settings, it's really common to take a part of the computation graph as a module and take input from different sources. It turns out that Tensorboard does not recognize that the same part of the graph is used for different inputs. Thus it fails to use a common scope for them.

**Comment by axelv** [comment]
Reusing a lot of layers I encounter the same messy graphs in TensorBoard.
It would be nice to have the two namespaces merged if they contain the same variables!

**Comment by LiJinyao** [comment]
Have same issue

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by axelv** [comment]
Any updates on this in recent releases?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by sleighsoft** [comment]
Same issue here

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tatatodd** [comment]
Maybe @jart has some ideas?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by sleighsoft** [comment]
A workaround to prevent variable scope from opening new namescopes is the following

```python
from tensorflow.python.ops import variable_scope as var_scope

def simple_variable_scope(name_or_scope, reuse=None):
  """Creates a variable scope without also creating a name scope."""
  return var_scope.variable_scope(name_or_scope, reuse=reuse,
                                  auxiliary_name_scope=False)
```

**Comment by angersson** [comment]
@dsmilkov Is this a tensorboard-side issue?

**Comment by sleighsoft** [comment]
I still observe that even when a variable scope does not create an extra name scope tensorboard still treats parts of the graph that receive varying inputs as multiple/duplicate nodes.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rajnunes** [comment]
Does this Problem by any chance take up excess gpu memory ?

**Comment by sleighsoft** [comment]
I would hope not, but feel free to create a test case.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by qmeeus** [comment]
I had the same issue issue: each time the variable_scope is closed and reopened, it creates a new scope name by appending some integer at the end. As a consequence, the variables are duplicated (as we can witness on tensorboard).
My (unelegant) solution to the problem: add everything that has to rest under the same scope in one go and close the scope afterward. I am aware that this might not be a solution in a lot of cases but I hope that it will help at least some!

**Comment by bhack** [comment]
Anybody tested if this has any memory impact?

**Comment by bhack** [comment]
@av8ramit sorry I didn't want to bother you. Was just for the @tensorflowbutler counting behavior :smile:  I think that the nagging assignee counter doesn't need to have a reset after non TF team members comments.

**Comment by sleighsoft** [comment]
@qmeeus If you really want to reuse name scopes then you can do this:
```python
def absolute_name_scope(scope):
  """Builds an absolute tf.name_scope relative to the current_scope.
  This is helpful to reuse nested name scopes.

  E.g. The following will happen when using regular tf.name_scope:

    with tf.name_scope('outer'):
      with tf.name_scope('inner'):
        print(tf.constant(1)) # Will print outer/inner/Const:0
    with tf.name_scope('outer'):
      with tf.name_scope('inner'):
        print(tf.constant(1)) # Will print outer/inner_1/Const:0

  With absolute_name_scope:

    with absolute_name_scope('outer'):
      with absolute_name_scope('inner'):
        print(tf.constant(1)) # Will print outer/inner/Const:0
    with absolute_name_scope('outer'):
      with absolute_name_scope('inner'):
        print(tf.constant(1)) # Will print outer/inner/Const_1:0
  """
  current_scope = tf.get_default_graph().get_name_scope()
  if not current_scope:
    if scope.endswith('/'):
      name_scope = tf.name_scope(scope)
    else:
      name_scope = tf.name_scope('{}/'.format(scope))
  else:
    name_scope = tf.name_scope('{}/{}/'.format(current_scope, scope))
  return name_scope
```

Tensorflow team, I can also add this as a PR if desired. I find this rather useful to get clean scopes in Tensorboard.

**Comment by av8ramit** [comment]
It is not a problem at all @bhack and absolutely no bother. I'm always here to help the community. I'll default to Andy on this one.

**Comment by rajnunes** [comment]
Im not sure if it does have an impact on memory usage but I was training a model similar to segnet with image input size as 640x1024 and, at max I could only set my batch size to 1 image which is kind of strange (not sure if it should take that much memory) i was using a titanX (12gb ) and it ended up using the whole 12 gb.
(if I tried increasing batch size to anything greater than 1 I got CUDA out of memory error)
Is there any way to check how much memory a particular model should be taking ?

**Comment by bhack** [comment]
@rajnunes Have you tried to check with https://www.tensorflow.org/guide/graph_viz#runtime_statistics to check the memory?

**Comment by qmeeus** [comment]
thanks @sleighsoft, appreciate it! 

@bhack the memory use was the thing that tipped me off. Indeed, I think that the duplication impacts memory, as I was getting those "More than 10% of total memory allocated" warnings and eventually, the program crashed... The warning disappeared after I adapted my code to avoid scope duplication and I can now train my network without crashing my computer.

**Comment by ewilderj** [comment]
@bhack if there are systemic issues with the bug nagging/updating, then let's file this as a bug itself so we can track it, assign it, and collect the multiple instances of where it's happening. Thanks for highlighting this!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by josh11b** [comment]
The normal TensorFlow way to handle using the same scope in two places is:

```
    with tf.name_scope('outer'):
      with tf.name_scope('inner') as scope:
        print(tf.constant(1)) # Will print outer/inner/Const:0
    with tf.name_scope(scope):
      print(tf.constant(1)) # Will print outer/inner/Const_1:0
```

**Comment by nlgranger** [comment]
AFAIK, this bug is not about shared scopes but having variables abnormally duplicated in tensorboard plots when said variable is used at different locations in the code.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chihuahua: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gsujansai** [comment]
I am facing the same issue. Its a problem with tensor-board creating extra instances of shared layers.

`def inference():
    with tf.variable_scope("inner", reuse=tf.AUTO_REUSE):
        x = tf.placeholder(dtype = tf.float32, shape=(None, 30,30,3))
        h = tf.layers.conv2d(x, 5, [3, 3], [1, 1], padding='SAME')
        print(tf.get_default_graph().get_name_scope()) 
        return h

with tf.variable_scope("outer"):
    x = inference() _# will print outer/inner_
    x2 = inference() _# will print outer/inner_1_
`
But output of the model summary shows that the variables are being shared.

`outer/inner/conv2d/kernel:0 (float32_ref 3x3x3x5) [135, bytes: 540]
outer/inner/conv2d/bias:0 (float32_ref 5) [5, bytes: 20]
Total size of variables: 140
Total bytes of variables: 560`

@sleighsoft's PR is still in review. Is there any suggested workaround?

**Comment by jvishnuvardhan** [comment]
This is a stale issue. Please check the issue with latest TensorFlow. If the issue still persists in the newer version of TF, please feel free to reopen it by providing details about the issue and a standalone code to reproduce the issue. Thanks!

**Comment by cipollone** [comment]
The graph associated to this model:
```python
import tensorflow as tf

# Meaningless model
input_A = tf.keras.Input(shape=(1,2), name='Input_A')
input_B = tf.keras.Input(shape=(1,2), name='Input_B')

inner_block = tf.keras.Sequential(
    [ tf.keras.layers.Dense(1) for i in range(50) ])

output = inner_block(input_A) + inner_block(input_B)

model = tf.keras.Model((input_A,input_B), output)

# Write
tf.keras.callbacks.TensorBoard('.', write_graph=True).set_model(model)
```
is:
![Screenshot from 2020-01-05 17-13-11](https://user-images.githubusercontent.com/25563883/71782839-f2255700-2fde-11ea-9efd-3296200885fc.png)
All layers and namespaces are working as expected, however the data flow is made more complex by dependencies of ReadVariableOp inside sequential_1. In more complex models, the real inputs become the thinner connections and they are hard to follow. Also, shared layers appear with different colors or gray (probably due to the different ops inside the original and reused variables).

I don't want to force Tf to merge namescopes, as this would make each block accept multiple outputs, giving a wrong visual hint (multiple dependencies, not runs).
A nice toggle in TensorBoard to exclude inputs of ReadVariableOp (and why not, control dependencies) would simplify graphs a lot.

Details: Tensorflow 2.1.0-rc0 built from source.

**Comment by sleighsoft** [comment]
Maybe you should open this as a separate issue at /tensorboard.

**Comment by cipollone** [comment]
Posted [https://github.com/tensorflow/tensorboard/issues/3118](url), thanks
